Members

GMSoudersJr opened the pull request:
### Centered Content
![fCC Seoul Members Centered Content](https://github.com/Free-Code-Camp-Seoul/fcc-seoul-website/assets/44438849/e0647864-18e4-4e5f-a60e-f60f469edf72)
- Looks a bit off with only 5 members, but should look better as more members are added.


### Consistent Font-Size
![fCC Seoul Events Title Consistent font-size](https://github.com/Free-Code-Camp-Seoul/fcc-seoul-website/assets/44438849/55b6c79a-ff44-4a13-808e-46e627c481ca)
- Events page now has same font-size for its first element as the other pages.

GMSoudersJr committed:
fix(members): add heading tag

Change p to h1 since this is the title of the page.

GMSoudersJr committed:
feat(members): center content

Center the title and the member cards given space between.

GMSoudersJr committed:
feat(members): center member cards

Actually, center had better feel since all cards with have the same gap
between them.

GMSoudersJr committed:
feat(page headings): consistent font-size

Consistent font-sizes for the first element on each page.

GMSoudersJr committed:
fix(members): dynamic member list

There is a Google Sheets API for the member data. The page fetches the
data. It needs to be optimized, though.

GMSoudersJr commented:
I'm working on the dynamic list, but I am having trouble figuring out Suspense or how to deal with the lag of getting the list members.

GMSoudersJr committed:
perf(members): attempting to load members

Loading is slow so it needs an indicator for UX

GMSoudersJr commented:
The latest commit is my failed attempt at displaying a loading indicator.

GMSoudersJr commented:
I tried both `getStaticProps` and `getServerSideProps`. Both did the same thing as far as I could tell. I must have implemented something wrong.

ivanms1 commented:
could you try the following?
- use `getStaticProps`
- remove suspense
- `yarn build`
- `yarn start`

do you still see a loading?
it should be instant now

GMSoudersJr commented:
Thanks for the feedback.

### The issue I have
- Click on Members tab in the navbar.
- No indication that I clicked.
 - `<p>Loading...</p>` does not show.
- A few seconds later, the page loads with data from the spreadsheet.
- Every navigation to Members takes a few seconds without an indicator for the user.
- If the spreadsheet is updated, the next navigation to Members shows the update, but as always, it takes a few seconds.
- I figure it will always take a while to run through the spreadsheet and create the JSON for the the API.
- That's why I think an indicator of some type is necessary.

### With your suggestion
- The page loads **immediately** with data from the spreadsheet 👍 
- However, if the spreadsheet is changed, the website does not get the updated information. (It only gets the information at build time, correct?)

How often do we want the member list to update?

ivanms1 commented:
> However, if the spreadsheet is changed, the website does not get the updated information. ?

You are correct

In`getServerSideProps` the fetch of the members blocks the rendering of page until that call is done, that's why suspense didn't work.

It looks like we have a couple of options

1. Use `getStaticProps`. There is a `revalidate` option for getStaticProps but I am not sure if it works in fly.io. In the worst case scenario we would have redeploy often, maybe every Sunday?

2. Move the loading logic to `_app.tsx` and do something similar to [this](https://stackoverflow.com/a/60756105). I am not a big fan of this.

3. Move the member fetching logic to the client side (Members Page Component) and use something like react query or just plain fetch.

GMSoudersJr commented:
So when would Suspense actually work? I'm confused.


1. I think this is the best option. We can always test it. And like you said
> worst case scenario, we would have to redeploy often...
- I tested `revalidate` with a short interval on a local build and it worked fine 
2. Seems like a bit much. I am also not a fan of that.
3. Maybe a last resort.

GMSoudersJr commented:
Does fly.io stop revalidation?
How does the homepage stay current?

ivanms1 commented:
let's do option 1 then

> Does fly.io stop revalidation?

I am not sure, I can't find any sources stating that it does or it doesn't. I think we can just test it.

> How does the homepage stay current?

It's using `getServerSideProps`

GMSoudersJr commented:
`getServerSideProps` that's right. 바보야! I knew that

GMSoudersJr committed:
fix(members): member data loading immediately

Use getStaticProps with revalidate set to one week. We can see how it
work and then update accordingly.

GMSoudersJr committed:
feat(events): consistent font sizes

Make font-size in media-query for events the same as other pages

GMSoudersJr commented:
I shared the Google Sheet with you

ivanms1 commented:
Cool, I'll review this later today, we don't need the loading component anymore right?

GMSoudersJr committed:
chore(loading): remove unused loading component

GMSoudersJr commented:
Loading removed

ivanms1 commented on src/const/index.ts:
could we reduce this to one hour?

ivanms1 commented on src/services/members/index.ts:
[let's try to avoid type assertion when we can](https://mainawycliffe.dev/blog/type-assertions-in-typescript/)
```suggestion
export const getMemberInfo = async (): Promise<MemberData[]> => {
  const res = await fetch(MEMBERS_API);
  const responseJSON = await res.json();
  const memberData = responseJSON.data;

  return memberData;
};
```

ivanms1 reviewed (approved):
Great job 🎉 
Just some minor comments

GMSoudersJr commented on src/const/index.ts:
sure can

GMSoudersJr commented on src/services/members/index.ts:
Noted!

GMSoudersJr committed:
fix(members): type assertion and revalidate time

remove type assertion and shorten revalidate time.

GMSoudersJr commented:
Thanks. And thanks for the feedback
